src/clj/puppetlabs/puppetserver/jruby_request.clj

```diff
@@ -67,10 +67,14 @@
      (handler (assoc request :jruby-instance jruby-instance)))))
 
 (defn get-environment-from-request
-  "Gets the environment from a request."
+  "Gets the environment from the URL or query string of a request."
   [req]
-  (-> req
-      (get-in [:params "environment"])))
+  ;; :environment is the value of a path component, "environment" is the value
+  ;; of a query string parameter. If environment is an inherent part of the
+  ;; resource being requested (part of the path), we prefer that so it doesn't
+  ;; get overridden by a rogue query string parameter.
+  (or (get-in req [:params :environment])
+      (get-in req [:params "environment"])))
```

Comment:
Picking up this from the comment thread on the old commit so that it's less invisible.

@camlow325 said:

> stringify-keys just looks at the generic :params map, stringifying any keywords found for keys. I was initially just naiively thinking about instead having it walk the generic :params map and choose a keyword representation over a string one when both are present.
>
> I like your idea about merging the params from different sources better, though - stringifying keys for each as you go - with route-params winning over query-params, winning over form-params where there are matching keys. Maybe we should pull all of this up into some higher level middleware replacement for the params handling in wrap-params-for-jruby? If we did that, seems like we wouldn't even need the above code because we would have normalized to a single "environment" key being in the map at the point this code is being evaluated?

@camlow325 that sounds like a reasonable thing to consider but is probably more than what I was imagining for right now.  Right now I'm just thinking about a smaller change that would allow us to get this merged for @nicklewis and we could circle back around later to improve it.

The idea that I'm floating is just to modify these two lines of code so that the first one digs into the 'route-params' map to see if there was a path element for the environment, and then *maybe* also changing the second line here to specifically dig into the `query-params` map.  That last part might not be a good idea though since Puppet may send that "query param" up as part of the form-encoded POST body.  But I think the first part about specifically looking in `route-params` for the value that came out of the URL path should be ok?